apps/src/util/i18nStringTrackerWorker.js

```diff
@@ -0,0 +1,104 @@
+import $ from 'jquery';
+
+/**
+ * Gets the singleton instance of an I18nStringTrackerWorker
+ * @returns {I18nStringTrackerWorker}
+ */
+export function getI18nStringTrackerWorker() {
+  return new I18nStringTrackerWorker();
+}
+
+/**
+ * A singleton class which buffers i18n string usage data and sends it in batches to the '/i18n/track_string_usage' API.
+ */
+class I18nStringTrackerWorker {
+  constructor() {
+    // Check if there is already a singleton instance.
+    const instance = this.constructor.instance;
+    if (instance) {
+      return instance;
+    }
+
+    // Set the singleton instance to this instance if it doesn't exist already.
+    this.constructor.instance = this;
+
+    /**
+     * A buffer of records to be sent in batches. It should have the following structure:
+     * {
+     *   'http://studio.code.org/s/20-hour/stage/2/puzzle/4': {
+     *     'common_locale': [ 'curriculum', 'teacherForum', 'professionalLearning', ...]
+     *   }
+     * }
+     */
+    this.buffer = {};
+  }
+
+  /**
+   * Buffers the given i18n string usage data to be sent later.
+   * @param stringKey [String] The key used to look up the i18n string value e.g. 'curriculum'
+   * @param source [String] Context about the file i18n value was looked up in e.g. 'common_locale'
+   * @param url [String] The page being rendered when the i18n string was looked up.
+   */
+  log(stringKey, source, url) {
+    if (!stringKey || !source || !url) {
+      return;
+    }
+
+    this.buffer[url] = this.buffer[url] || {};
+    this.buffer[url][source] = this.buffer[url][source] || new Set();
+    this.buffer[url][source].add(stringKey);
+
+    // schedule a buffer flush if there isn't already one.
+    if (!this.pendingFlush) {
+      this.pendingFlush = setTimeout(
+        function() {
+          this.flush();
+        }.bind(this),
```

Comment:
I think you can just do

```suggestion
        () => this.flush(),
```

Reply:
nice!